File: DSA09015.py
from collections import defaultdict, deque

def function():
    v, e = map(int, input().split())
    ls = list(map(int, input().split()))

    while len(ls) < 2 * e:
        ls += list(map(int, input().split()))

    adj = defaultdict(list)
    for i in range(0, 2 * e, 2):
        adj[ls[i]].append(ls[i + 1])

    visited = defaultdict(bool)

    def dfs(bVertex):
        stack = deque()
        stack.append(bVertex)
        visited[bVertex] = True
        onstack = defaultdict(int)
        while stack:
            u = stack.pop()
            onstack[u] = False
            for x in adj[u]:
                if not visited[x]:
                    visited[x] = True
                    onstack[x] = True
                    onstack[u] = True
                    stack.append(u)
                    stack.append(x)
                    break
                elif onstack[x]:
                    return True
        return False

    for i in range(1, v + 1):
        if not visited[i]:
            if dfs(i):
                return True

    return False


if __name__ == "__main__":
    for i in range(int(input())):
        print("YES" if function() else "NO")

File: DSA09015.cpp
#include <iostream>
#include <vector>
#include <map>
#include <deque>

using namespace std;

bool function() {
    int v, e;
    cin >> v >> e;
    vector<int> ls;
    int temp;
    for (int i = 0; i < 2 * e; ++i) {
        cin >> temp;
        ls.push_back(temp);
    }

    map<int, vector<int>> adj;
    for (int i = 0; i < 2 * e; i += 2) {
        adj[ls[i]].push_back(ls[i + 1]);
    }

    map<int, bool> visited;
    map<int, int> parent;
    deque<int> stack;

    auto dfs = [&](int bVertex){
        stack.push_back(bVertex);
        bool has_cycle = false;

        while (!stack.empty()) {
            int u = stack.back();
            stack.pop_back();
            if (!visited[u]) {
                visited[u] = true;
                for (int x : adj[u]) {
                    if (!visited[x]) {
                        stack.push_back(x);
                        parent[x] = u;
                        break;
                    } else if (parent[u] != x) {
                        has_cycle = true;
                        break;
                    }
                }
            }
            if (has_cycle) break;
        }
        return has_cycle;
    };

    for (int i = 1; i <= v; ++i) {
        if (!visited[i]) {
            if (dfs(i)) {
                return true;
            }
        }
    }
    return false;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        cout << (function() ? "YES" : "NO") << endl;
    }
    return 0;
}
